### DISCUSSION TITLE
Error Spawning fortls

### DISCUSSION THREAD

**Discussion by AlaBahrami** [fortran-lang/fortls#282] [Q&A] [answered]
Hi,
I have this error while I am running VScode:
Error Spawning fortls: Please check that fortran-language-server is installed and in your path. 
Source: FORTRAN IntelliSense (Exyension). 
I pip installed fortls via Ubuntu (WSL) of VSCode’s Terminal and added both fortls.path and  executablePath. 

Package versions:
VSCode: 1.78.2 
Fortls: 2.13.0
Modern Fortran: 3.2.0

Would you tell me how can I fix this issue?
Thanks

**Comment by gnikit** [comment, marked as answer]
Hey @AlaBahrami  [Modern Fortran](https://marketplace.visualstudio.com/items?itemName=fortran-lang.linter-gfortran) and FORTRAN Intellisense are not compatible. The same goes for [fortls](https://fortls.fortran-lang.org/) and FORTRAN Intellisense.
We can only guarantee that our language server (fortls) works with our VS Code extension (Modern Fortran).

I hope that clears things up a bit.

**Comment by AlaBahrami** [comment]
Hi @gnikit. Thanks for your response. Yes, I see the warning message pops out that the FORTRAN Intellisense is not compatible with Modern Fortran. So, I disabled the FORTRAN Intellisense extension. It is not obvious to me why in the FORTRAN Intellisense page (https://github.com/hansec/vscode-fortran-ls) it is declared the extension requires Modern Fortran or Fortran. 

However, I still have problem related IDE. It seems it is not active and features are not active. The error message is : 
Modern Fortran Error starting fortls: Check that fortls is in your PATH or that "fortran.fortls.path" is pointing to a fortls binary.

**Reply by gnikit** [reply]
Well, FORTRAN Intellisense as you can see is pretty-much dead (the principle author stopped contributing 3y ago). Regardless of that, the extension was not doing anything special, other than informing VSCode that a language server (fortls or fortran-language-server) can be registered. This is done natively in Modern Fortran now, without any need for 3rd party extensions (one stop shop).

If `fortls` cannot be detected that usually means that is not part of your systems's `$PATH`. I am working on a mechanism that tries a few more locations (depending on the OS) before it raises a warning. What you can do in the meantime is define in your settings manually, the absolute path to fortls

```json
  "fortran.fortls.path": "/home/gn/.local/bin/fortls", 
```

> NOTE: Rember on Windows the binary is called `fortls.exe`

You can find  where pip installs fortls via

```sh
pip show --files fortls
````

**Comment by AlaBahrami** [comment]
Thanks @gnikit for your explanation. Yes, it seems the FORTRAN Intellisense has not been updated since three years ago. 

I still have the issue regarding the detection of `PATH`. When I use 

`pip show --files fortls `

It shows me the directory of Python packages. It doesn't direct me to the `bin` . I can find the `bin` directory by going back some directories from the `python3.8/site-packages` or use `which fortls` will give me the exact bin directory for fortls. Here I don't see the `fortls` has the extension '.exe'. 

I set absolute path as the following, but still the IDE is not working and I am getting the same error as before. 

`"fortran.fortls.path": "D:/pdf/yes/bin/fortls"`

I tried on other workstation with different directory and I have the same issue. For both setting of `PATH` does not work.

**Reply by gnikit** [reply]
I think there is some confusion here, `which` is a UNIX shell command, if you are on Windows and are using which that probably means that you are using something else, other than powershell or cmd. The equivalent windows command is `where`.

Aside from that, VS Code on Windows (AFAIK) runs on the OS level and hence does not inherit variables set locally in a shell.
Hence the PATH discovery needs to be global.

If you have specified `"fortran.fortls.path"`, then it will try and spawn that binary at the OS level^. I think Windows PATHs use the `\\` seperator and extensions installed on windows via pip get installed in `..\Scripts\`, not `bin` and end with `.exe`.
There is nothing stopping you from passing whatever binary in the VS Code option, with whatever extension, VS Code Modern Fortran will simply grab that and attempt to run it a subprocess.
The easiest way to make sure you `fortls` installation works is run the application itself from a terminal against a Fortran source file e.g.

```sh
Path/to/fortls --debug_filename myfile.f90 --debug_parser
```

For any additional help, please install the Pre-Release, set the Logging to Debug and post the logs in the OUTPUT pane.

----

^: more or less, it's actually more complicated than that, but it doesn't affect the user.

**Comment by AlaBahrami** [comment]
Hi @gnikit. Yes, I am using Ubuntu as I explained before and I compile and run a Fortran code in Linux via Ubuntu WSL. I have never run a code in the Windows environment before using Powershell or cmd. I became confused also why I install `fortls` via `Ubuntu(wsl)` and specified the directory of installation in the VS Code does not recognize it. So, from your explanation, I found VS Code does not recognize the local variables installed in Ubuntu, Anaconda, or others. Since I have not installed Python on my machine and I use it from Anaconda I follow these steps to recognize `fortls` in Windows:

1.	Install Python version 3.11 and add it to the System Variables 
2.	`Python -m pip install fortls` via using cmd
3.	Use `where` to find the location of installed fortls which the same directory as Python has been installed in the `Scripts`
4.	Specify the `"fortran.fortls.path"` in the `settings.json` file

Finally, I can use fortls and its features are active. For example, I use `peek` and it works. I tesed on two of my systems and they are working.
If you can elaborate on the explanation and workflow of installation of fortls for the Windows system that will be great. I see a lot of problems reported on web, but the workflow is not clear. The problem I see here is that VS Code does not recognize the local installation of fortls in other command lines using Ubuntu and you got confused why even though the package has been installed, the VS Code does not detect it. You can also try to find a mechanism to recognize installed fortls from Ubuntu(wsl) in the VS Code.

**Reply by gnikit** [reply]
> If you can elaborate on the explanation and workflow of installation of fortls for the Windows system that will be great. I see a lot of problems reported on web, but the workflow is not clear

fortls needs to be available in the OS's `$PATH`. That's what the VS Code extension has access to.

I haven't found a way to reliably propagate environmental variables from a set Python interpreter or from a development environment all the way down to the Modern Fortran extension. If you have any ideas, or stumble onto a VS Code extension that can do that, feel free to open a Pull Request or a Discussion over at https://github.com/fortran-lang/vscode-fortran-support.


I will add it on my TODO list to try and understand (and document) what happens when you use a development environment like WSL.
I was under the impression that VS Code run inside WSL, which served its results to the VSCode instance installed at the OS level, which was acting purely as a client.

If that is the case, based on my own experiences with Remote Server environments in VSCode, you only have to setup your extensions server side. So you would need to install fortls, and add its pip dir to `$PATH` or point `fortran.fortls.path` to that location inside WSL. From what you described that is not the case.
Maybe WSL is somehow different, I will have to check.

**Reply by AlaBahrami** [reply]
From my understanding, the VS Code extension does not have access to WSL development environmenet that's why VS Code does not recognize the installed fortls inside WSC. 

> If you have any ideas, or stumble onto a VS Code extension that can do that, feel free to open a Pull Request or a Discussion over

I don't have any ideas at this moment how to do propogate the environemnt variables to the Modern Fortran extension at this moment, but it should be a solution for this
